File: arrays_and_hashing/36_valid_sudoku.py
# https://leetcode.com/problems/valid-sudoku/

from collections import defaultdict


class Solution:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        rows = defaultdict(set)
        cols = defaultdict(set)
        grid = defaultdict(set)

        for row in range(9):
            for col in range(9):
                val = board[row][col]
                if val == '.':
                    continue

                if (
                    val in rows[row] or
                    val in cols[col] or
                    val in grid[(row // 3, col // 3)]
                ):
                    return False

                rows[row].add(val)
                cols[col].add(val)
                grid[(row // 3, col // 3)].add(val)

        return True

File: go/arrays_and_hashing/36_valid_sudoku.go
package arraysandhashing

func IsValidSudoku(board [][]byte) bool {
	var rows [9][9]bool
	var cols [9][9]bool
	var grid [9][9]bool

	for row := range board {
		for col, val := range board[row] {
			if val == '.' {
				continue
			}
			index := val - '1'
			gridKey := row/3 + col/3*3
			if rows[row][index] || cols[col][index] || grid[gridKey][index] {
				return false
			}
			rows[row][index] = true
			cols[col][index] = true
			grid[gridKey][index] = true
		}
	}
	return true
}
